File: lib/controllers/order_controller.dart
import 'dart:convert';
import 'package:flutter/material.dart';
import 'package:get/get.dart';
import 'package:http/http.dart' as http;
import 'package:utibu_health/controllers/auth_controller.dart';
import 'package:utibu_health/controllers/drug_controller.dart';
import 'package:utibu_health/models/drug_model.dart';
import 'package:utibu_health/models/order_model.dart';
import 'package:utibu_health/models/user_model.dart';
import 'package:utibu_health/utils/colors.dart';
import 'package:utibu_health/utils/strings.dart';

class OrderController extends GetxController {
  late int userId;

  // OrderController(this.currentUserId); // Constructor to accept current user ID

  // int currentUserId; // Variable to hold current user ID
  User? currentUser;

  // void setCurrentUserId(int userId) {
  //   currentUserId = userId;
  // }

  // Instance of drugController
  final DrugController drugController = Get.find();
  final AuthController authController = Get.find();
  // List to hold available medications
  RxList<Drug> availableMedications = <Drug>[].obs;

  @override
  void onInit() {
    super.onInit();
    fetchDrugs(); // Call fetchDrugs on initialization
  }

  void fetchDrugs() {
    drugController.fetchDrugs(); // Call fetchDrugs from drugController

    // Listen for changes in drugs list from drugController
    ever(drugController.drugs, (List<Drug>? drugs) {
      if (drugs != null) {
        // Update availableMedications list with fetched drugs
        availableMedications.assignAll(drugs);
        // Refresh the list to notify any listeners
        availableMedications.refresh();
        // print(availableMedications);
      }
    });
  }

  final RxList<Order> orders = <Order>[].obs;
  RxString medicationName = ''.obs;
  RxInt quantity = 0.obs;
  int medicationId = 0;
  RxBool isOrderSuccessful = false.obs;
  var payment = 'Pay Later'.obs;
  var transactionCode = ''.obs;
  RxList<OrderItem> orderItems = <OrderItem>[].obs;
  RxDouble totalCalculatedAmount = 0.0.obs;
  double intermediateAmount = 0.0;
  var calculatedAmount = 0.0.obs;

  // Calculate the total amount based on order items and drug prices
  double calculateAmount() {
    double total = 0;
    for (var item in orderItems) {
      // Fetch drug based on ID from availableMedications
      Drug? selectedDrug =
          availableMedications.firstWhereOrNull((drug) => drug.id == item.id);
      if (selectedDrug != null) {
        total += item.quantity * selectedDrug.pricePerUnit;
      }
    }

    print(total.toString());
    return total;
  }

  // Add order item to the list
  void addOrderItem() {
    if (medicationName.isNotEmpty && quantity.value > 0) {
      // Fetch drug details based on ID
      Drug? selectedDrug = availableMedications
          .firstWhereOrNull((drug) => drug.id == medicationId);
      if (selectedDrug != null) {
        orderItems.add(OrderItem(
          medicationName: selectedDrug.medicationName,
          quantity: quantity.value,
          id: medicationId,
        ));
        intermediateAmount = calculateAmount();

        calculatedAmount.value =
            intermediateAmount.obs.value; // Update observable variable
        // Clear input fields
        medicationName.value = '';
        quantity.value = 0;

        Get.snackbar('Success', 'Item added to order successfully',
            backgroundColor: AppColors.kijaniColor, colorText: Colors.white);
      } else {
        // Show error snackbar if drug details not found
        Get.snackbar('Error', 'Drug details not found',
            backgroundColor: Colors.red, colorText: Colors.white);
      }
    } else {
      // Show error snackbar if fields are empty or invalid
      Get.snackbar('Error', 'Please enter valid medication and quantity',
          backgroundColor: Colors.red, colorText: Colors.white);
    }
  }

  // Setters for observable variables
  void setMedicationName(String value) => medicationName.value = value;
  void setQuantity(int value) => quantity.value = value;
  void setPayment(String value) => payment.value = value;
  void setTransactionCode(String value) => transactionCode.value = value;
  void setMedicationId(int value) => medicationId = value;

  Future<void> placeOrder() async {
    currentUser = authController.currentUser.value;

    try {
      // Bundle order information
      Order order = Order(
        orderItems: orderItems.toList(),
        userId: currentUser!.id,
        payment: payment.value,
        transactionCode: transactionCode.value,
        amount: calculateAmount(),
      );
      // print(order);
      // Send order request to the server
      // Construct the headers with the JWT token
      var headers = {
        'Content-Type': 'application/json',
        // 'Authorization': currentUser!.token, // Include the JWT token here
      };

      // Send order request to the server with headers
      var response = await http.post(
        Uri.parse('$baseUrl/api/orders/placeOrder'),
        body: jsonEncode(order.toJson()),
        headers: headers, // Include headers in the request
      );

      if (response.statusCode == 201) {
        var data = jsonDecode(response.body);
        if (data['success'] = true) {
          isOrderSuccessful.value = true;
          // Reset order details
          medicationName.value = '';
          quantity.value = 0;
        } else {
          isOrderSuccessful.value = false;
        }
      } else {
        throw Exception('Failed to place order');
      }
    } catch (e) {
      print('Error placing order: $e');
      isOrderSuccessful.value = false;
    }
  }

  // Remove order item from the list
  void removeOrderItem(int index) {
    if (index >= 0 && index < orderItems.length) {
      orderItems.removeAt(index);
      intermediateAmount = calculateAmount();

      calculatedAmount.value =
          intermediateAmount.obs.value; // Update observable variable
    }
  }
}

File: controllers/orderController.js
const Order = require('../models/Order');
const Statement = require('../models/Statement');
const Drug = require('../models/Drug');
const Payment = require('../models/Payment');
const sequelize = require('../config/index');


const placeOrder = async (req, res) => {
  const { orderItems, userId, payment, amount } = req.body;

  let transaction;
  try {
    // Start a transaction
    transaction = await sequelize.transaction();

    // Create the order
    const order = await Order.create({ userId: userId,
       medicationName:  orderItems[0].medicationName,
      quantity: orderItems[0].quantity,
      amount: amount, // Calculate order amount based on items
      payment,
    }, { transaction }); 

    // Check if payment is "Pay Now"


    if (payment === 'Pay Now') {
    
      const paymentRecord = await Payment.findOne({
        where: {
          amount: req.body.amount,
          transactionCode: req.body.transactionCode,
        
          // Add more conditions if needed
        },
        // raw: true
      });

    
      console.log('Payment Record:', paymentRecord);
      
      if (paymentRecord) {
        // console.log('Payment Record Found:', paymentRecord.toJSON());
        // Update payment status to Paid
        await order.update({ paymentStatus: 'Paid' }, { transaction });
      } else {
        console.log('No Payment Record Found');
      }      
    }
    // Create statements for each order item
    for (const item of orderItems) {
      const drug = await Drug.findByPk(item.id, { transaction });
      if (!drug || drug.quantity < item.quantity) {
        
        res.status(400).json({ status: false, message: `Drug ${item.medicationName} is not available in sufficient quantity` });
      }

      // Create the corresponding statement
      const statementData = {
        date: new Date(),
        description: `Ordered ${item.quantity} units of ${item.medicationName}`,
        amount: drug.price_per_unit * item.quantity,
        orderId: order.id,
        userId, // Use the userId from req.body
        quantity: item.quantity,
        medicationName: item.medicationName,
        paymentStatus: order.paymentStatus
      };
      await Statement.create(statementData, { transaction });

      // Update the drug quantity
      drug.quantity -= item.quantity;
      await drug.save({ transaction });
    }

    // Commit the transaction
    await transaction.commit();

    res.status(201).json({ status: true, message: 'Order placed successfully', order });
  } catch (error) {
    // Rollback the transaction if any error occurs
    if (transaction) await transaction.rollback();
    // console.log(req.body);
    console.error('Error placing order:', error);
    res.status(500).json({ error: 'Failed to place order' });
    }

};
// Function to calculate order amount based on items

module.exports = { placeOrder };
